File: android/src/main/java/app/plink/android/protocol/PlinkMessage.kt
package app.plink.android.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction

@Serializable
data class PlinkEnvelope(
    val version: Int = 1,
    val id: String,
    val type: String,
    val sentAt: String,
    val sourceDeviceId: String,
    val targetDeviceId: String,
    val requiresAck: Boolean = false,
    val payload: JsonObject
) {
    fun encode(): String = json.encodeToString(serializer(), this)

    companion object {
        private val json = Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
            prettyPrint = false
        }

        fun decode(raw: String): PlinkEnvelope {
            ReconnectPayloadPolicy.validateRawJSON(raw)
            ScreenPreviewPayloadPolicy.validateRawJSON(raw)
            FileTransferPayloadPolicy.validateRawJSON(raw)
            return decodeUnchecked(raw)
        }

        fun decode(raw: ByteArray): PlinkEnvelope = decode(decodeUtf8(raw))

        internal fun decodeUtf8(raw: ByteArray): String = Charsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(raw))
            .toString()

        internal fun decodeUnchecked(raw: String): PlinkEnvelope =
            json.decodeFromString(serializer(), raw)
    }
}

object PlinkEventType {
    const val PairingOffer = "pairing.offer"
    const val PairingConfirm = "pairing.confirm"
    const val DeviceStatus = "device.status"
    const val CallRinging = "call.ringing"
    const val CallEnded = "call.ended"
    const val MessageReceived = "message.received"
    const val MessageReply = "message.reply"
    const val ClipboardUpdated = "clipboard.updated"
    const val FileOffer = "file.offer"
    const val FileAccept = "file.accept"
    const val FileChunk = "file.chunk"
    const val FileProgress = "file.progress"
    const val FileComplete = "file.complete"
    const val FileResult = "file.result"
    const val FileCancel = "file.cancel"
    const val ScreenRequest = "screen.request"
    const val ScreenState = "screen.state"
    const val ScreenPull = "screen.pull"
    const val ScreenFrame = "screen.frame"
    const val ScreenIdle = "screen.idle"
    const val ScreenStop = "screen.stop"
    const val WebOpen = "web.open"
    const val MediaState = "media.state"
    const val MediaCommand = "media.command"
    const val PermissionState = "permission.state"
    const val Ack = "ack"
    const val Error = "error"
    const val ReconnectHello = "reconnect.hello"
    const val ReconnectChallenge = "reconnect.challenge"
    const val ReconnectProof = "reconnect.proof"
    const val ReconnectReverse = "reconnect.reverse"
    const val ReconnectReverseProof = "reconnect.reverse_proof"
    const val ReconnectReady = "reconnect.ready"
    const val ReconnectCommit = "reconnect.commit"
    const val ReconnectDone = "reconnect.done"
}

File: macos/Sources/PlinkCore/PlinkMessage.swift
import Foundation

public struct PlinkEnvelope: Codable, Equatable, Sendable {
    public var version: Int
    public var id: String
    public var type: EventType
    public var sentAt: Date
    public var sourceDeviceId: String
    public var targetDeviceId: String
    public var requiresAck: Bool
    public var payload: [String: PayloadValue]

    /// Use for wire input so event-specific raw JSON policies run alongside typed decoding.
    public static func decode(
        _ data: Data,
        reconnectValidation: ReconnectValidationPolicy = .production
    ) throws -> PlinkEnvelope {
        let envelope = try PlinkJSON.decoder().decode(PlinkEnvelope.self, from: data)
        if ReconnectPayloadPolicy.eventTypes.contains(envelope.type) {
            try ReconnectPayloadPolicy.validateRawJSON(data, validation: reconnectValidation)
        }
        try FileTransferPayloadPolicy.validateRawJSON(data)
        try ScreenPreviewPayloadPolicy.validateRawJSON(data)
        try ReconnectPayloadPolicy.validate(envelope, validation: reconnectValidation, plaintextBytes: data.count)
        return envelope
    }

    public init(
        version: Int = 1,
        id: String,
        type: EventType,
        sentAt: Date,
        sourceDeviceId: String,
        targetDeviceId: String,
        requiresAck: Bool = false,
        payload: [String: PayloadValue]
    ) {
        self.version = version
        self.id = id
        self.type = type
        self.sentAt = sentAt
        self.sourceDeviceId = sourceDeviceId
        self.targetDeviceId = targetDeviceId
        self.requiresAck = requiresAck
        self.payload = payload
    }
}

public enum PayloadValue: Codable, Equatable, Sendable {
    case string(String)
    case bool(Bool)
    case int(Int)
    case double(Double)

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Int.self) {
            self = .int(value)
        } else if let value = try? container.decode(Double.self) {
            self = .double(value)
        } else {
            self = .string(try container.decode(String.self))
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value):
            try container.encode(value)
        case .bool(let value):
            try container.encode(value)
        case .int(let value):
            try container.encode(value)
        case .double(let value):
            try container.encode(value)
        }
    }

    public var stringValue: String? {
        if case .string(let value) = self { return value }
        return nil
    }

    public var boolValue: Bool? {
        if case .bool(let value) = self { return value }
        return nil
    }

    public var intValue: Int? {
        if case .int(let value) = self { return value }
        return nil
    }
}

public enum EventType: String, Codable, Hashable, Sendable {
    case pairingOffer = "pairing.offer"
    case pairingConfirm = "pairing.confirm"
    case deviceStatus = "device.status"
    case callRinging = "call.ringing"
    case callEnded = "call.ended"
    case messageReceived = "message.received"
    case messageReply = "message.reply"
    case clipboardUpdated = "clipboard.updated"
    case fileOffer = "file.offer"
    case fileAccept = "file.accept"
    case fileChunk = "file.chunk"
    case fileProgress = "file.progress"
    case fileComplete = "file.complete"
    case fileResult = "file.result"
    case fileCancel = "file.cancel"
    case screenRequest = "screen.request"
    case screenState = "screen.state"
    case screenPull = "screen.pull"
    case screenFrame = "screen.frame"
    case screenIdle = "screen.idle"
    case screenStop = "screen.stop"
    case webOpen = "web.open"
    case mediaState = "media.state"
    case mediaCommand = "media.command"
    case permissionState = "permission.state"
    case reconnectHello = "reconnect.hello"
    case reconnectChallenge = "reconnect.challenge"
    case reconnectProof = "reconnect.proof"
    case reconnectReverse = "reconnect.reverse"
    case reconnectReverseProof = "reconnect.reverse_proof"
    case reconnectReady = "reconnect.ready"
    case reconnectCommit = "reconnect.commit"
    case reconnectDone = "reconnect.done"
    case ack
    case error
}
